Token grace period

dpsutton opened the pull request:
Add a grace period to token checks

The gist: create a grace period by which we allow previous token responses to remain in effect when the token checks error.
To achieve this, a new protocol GracePeriod with functions `save! token repsonse` and `retrieve token`. Current implementation is on top of a guava cache that lets us expire keys at 36 hours. It's completely up to the grace period for when retrieve works; the caller doesn't care, it just asks if there is a value in the grace period.

There's a function `decode-token` that will decode an airgapped token or hit our token check service with the token. This function is memoized at 12 hours. This function must return a response `{:valid ... :features}` or throw an error if its unable to decode the token.

Ther function `harness` invokes the above decode token, catches errors, hooks up the rate limiting, saves successful responses into the grace period and provides the grace period value in the case that the `decode-token` function errors.

```
 (*token-features*) ;; call this function to simulate the check of features
Checking with the MetaStore to see whether token 'f325...d01a' is valid...
Reporting Metabase stats: {... user counts etc}
 -> #{"features come back"}
 ;; meanwhile clear cache to simulate elapsing of 12 hours

 (*token-features*) ;; call function again to simulate check of features
2025-10-22 21:30:26,620 INFO premium-features.token-check :: Checking with the MetaStore to see whether token 'f325...d01a' is valid...
2025-10-22 21:30:26,622 ERROR premium-features.token-check :: Error fetching token status from https://token-check.metabase.com:
clojure.lang.ExceptionInfo: network issues ;; hit network issue

2025-10-22 21:30:26,632 INFO premium-features.token-check :: Using token from grace period
->> #{graceperiod network features}
...

;; here after 5 seconds (in real life will be 36 hours)
2025-10-22 21:31:00,064 WARN premium-features.token-check :: Removing token: f325...d01a from grace period cache {mb-quartz-job-type=Cache}
```

Everything else is trying to remove a bunch of complexity down to a single function that takes a token and returns information, clear area where all the caching, retry strategies, and grace periods exist in one lexical scope.


```clojure
metabase.premium-features.token-check=> (-token-status)
{:valid true, :status "Token is valid." :features [...]}
metabase.premium-features.token-check=> ;; turned off wifi
metabase.premium-features.token-check=> (clear-cache!) ;; clears the ttl, simulating expiration
nil
metabase.premium-features.token-check=> (-token-status)
{:valid true, :status "Token is valid." :features [...]}
metabase.premium-features.token-check=> ;; reenabled wifi
metabase.premium-features.token-check=> (-token-status)
{:valid true, :status "Token is valid." :features [...]}
```

And associated logs at the time:

```
;; intial real check
INFO premium-features.token-check :: Checking with the MetaStore to see whether token 'f325...d01a' is valid...
INFO premium-features.token-check :: Reporting Metabase stats: {...}
INFO premium-features.token-check :: Checking with the MetaStore to see whether token 'f325...d01a' is valid...
;; when wifi is off we hit the grace period
INFO premium-features.token-check :: Reporting Metabase stats: {...}
INFO premium-features.token-check :: Using token from grace period
;; wifi is reenabled we try again to hit store
INFO premium-features.token-check :: Checking with the MetaStore to see whether token 'f325...d01a' is valid...
INFO premium-features.token-check :: Reporting Metabase stats: {...}
```

dpsutton committed:
Add a grace period to token checks

There's more work on this namespace to be done i think, but this is a
first cut at it. There's a lot of control flow by exceptions which is
how we jump around in here. Fundamentally this should be far
simpler. Will take a crack at that shortly. I also don't love that this
just hammers the `fetch-token-status` function and then memoizes inside
of that. Would prefer simpler layout of `token-features` function and
that can be aware of population issues.

Adds a simple protocol: GracePeriod, which has two functions: save! and
retrieve. On successful token check, we `save!` the response. On errors,
we `retrieve` previous responses. This obviously can be nil in the case
we have never had a successful response, and will become nil once the
grace period's internal notion of time expires.

This implementation is based on guava's cache which has a nifty
expireafterwrite functionality.

```
 (*token-features*) ;; call this function to simulate the check of features
Checking with the MetaStore to see whether token 'f325...d01a' is valid...
Reporting Metabase stats: {... user counts etc}
 -> #{"features come back"}
 ;; meanwhile clear cache to simulate elapsing of 12 hours

 (*token-features*) ;; call function again to simulate check of features
2025-10-22 21:30:26,620 INFO premium-features.token-check :: Checking with the MetaStore to see whether token 'f325...d01a' is valid...
2025-10-22 21:30:26,622 ERROR premium-features.token-check :: Error fetching token status from https://token-check.metabase.com:
clojure.lang.ExceptionInfo: network issues ;; hit network issue

2025-10-22 21:30:26,632 INFO premium-features.token-check :: Using token from grace period
->> #{graceperiod network features}
...

;; here after 5 seconds (in real life will be 36 hours)
2025-10-22 21:31:00,064 WARN premium-features.token-check :: Removing token: f325...d01a from grace period cache {mb-quartz-job-type=Cache}
```

dpsutton committed:
tests that updating value populates grace period cache

dpsutton committed:
couple changes

protect against storing nil value. guava cache doesn't like it so gone
it is.

added some tests:

```clojure
(testing "Setting a new value populates the grace period cache"
    (let [new-token (tu/random-token)]
      (binding [http/request (fn [& _] {:status 200
                                        :body "{\"valid\":true,\"status\":\"fake\",\"features\":[\"fake\",\"features\"]}"})]
        (mt/discard-setting-changes [premium-embedding-token]
          (premium-features.settings/premium-embedding-token! new-token)
          (is (= #{"fake" "features"} (token-check/*token-features*)))
          (is (= {:valid true :status "fake" :features ["fake" "features"]}
                 (token-check/retrieve token-check/grace-period new-token)))))))
```

But this ended up leaving a setting table value for this value. i don't
feel like looking into discard setting changes so i'm gonna just remove it

dpsutton committed:
attempt to simplify the api a bit

the thee important functions are now `token-information`, `harness` and
`decode-token`.

Everyone should use token-information. Harness hooks up the grace period
and retry strategies, and `decode-token` only cares about decoding a
token. It should return the token-status map {:valid true ...} or throw.

dpsutton committed:
fixup test

dpsutton committed:
clean up tests and rename a bit

dpsutton committed:
remove comment block

dpsutton committed:
remove commented out grace period

ericnormand commented on src/metabase/premium_features/token_check.clj:
```diff
@@ -55,10 +59,6 @@
              (str/replace  #"/$" "")))
    "https://token-check.metabase.com"))
 
-(def store-url
```
Is this fallback url not used anymore? Is the grace period a replacement for a fallback?

ericnormand commented:
I also don't understand why there's a memoization for 12 hours and a "grace period" for 36 hours. What's the idea there? Do they cache different things? Why not memoize for 36 hours?

dpsutton commented:
Great questions @ericnormand !

- positive vs negative results:  By positive i mean the token check server answered whether your token is valid or not. A 200 response with features is good, but so is a 404 saying it doesn't know that token. Both are authoritative responses about your token.

```
❯ http get https://token-check.staging.metabase.com/api/notachancethisisavalidtoken/v2/status
HTTP/1.1 404 Not Found
Content-Type: application/json;charset=utf-8
{
    "status": "Token does not exist.",
    "valid": false
}
```

You are right that we cache this, but we _should_. Your token is invalid.

- negative results: these are responses that aren't about your token. Think network issues, server 500's, etc. It's not authoritative about your token. We do not want to save this. It is transient and not related to your token.

Let me explain a bit more about the happy path here. We memoize the `decode-token` function. We call that repeatedly but rely on the memoization to provide the features for 12 hours. After that, we hit the license check server again. If this succeeds, you are good for another 12 hours, at which point we will check again.

Our desired behavior is "check the license server every 12 hours. If there's an issue reaching that service, let the last good value stand in place and not immediately remove all features. But only up to a point. At some point we have to remove the features." This PR introduces the grace period that if license checks start failing, we use the last known good value for up to 24 hours grace period. A token check is good for 12 hours, we hold onto the value for 36 hours: which means if we can't hit the license server, we have 24 hours of allowing the previously known features to persist.

ericnormand commented on src/metabase/premium_features/token_check.clj:
```diff
@@ -284,42 +222,124 @@
-  "Getter for the [[metabase.premium-features.settings/token-status]] setting."
-  []
-  (some-> (premium-features.settings/premium-embedding-token) (fetch-token-status)))
+        (let [f' (fn [token]
```
Can we define this at the top level and make it private?

galdre commented on src/metabase/premium_features/token_check.clj:
```diff
@@ -284,42 +222,124 @@
+                                               (log/warnf "Removing token: %s from grace period cache"
+                                                          (u.str/mask (.getKey rn)))))))
+                        (build))]
+    (reify GracePeriod
```
I appreciate the protocol. The rest of the code doesn't need to know or care that it's a guava-cache specifically, and you could easily mock out a different impl without touching the rest of the file.

The name `GracePeriod` feels a little odd to me, more the name of the _why_ than the _what_, but that's a nitpick and I don't currently have a better idea.

dpsutton committed:
bit of cleanup

dpsutton committed:
fix with-random-premium-token! macro

dpsutton committed:
function composes them instead of caller

rather than a caller needing to know all of the layers, i thread them
for you and let you provide all arugments in one function call.

dynamic var for needing to customize layers in tests. largely for
removing the circuit-breaker which is stateful and hard to restart if
you want to simulate errors in tests

dpsutton commented on src/metabase/premium_features/token_check.clj:
```diff
@@ -286,40 +218,190 @@
+                                   :delay-ms                          (u/seconds->ms 30)
+                                   :success-threshold                 1}
+                 :timeout-ms      (u/seconds->ms 10)
+                 :ttl-ms          (u/hours->ms 12)
```
i'm on the fence about this. I think i like the cache-ttl-ms but that's kinda redundant to ttl-ms. The timeout-ms could conceivably go into the regular token check thing and not be part of the circuit breaker so it feels a bit like implementation detail getting into the exposed api

dpsutton committed:
rename test
